### PR TITLE
MSPSDS-938 Business type selection on case page - add business

### PR DESCRIPTION
**Pull request by samlionheart**
Initially used params to pass through the business type but couldn't return errors, so turned it into a wicked form

### CODE REVIEW

File: mspsds-web/app/views/investigations/businesses/type.html.slim [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  h2.govuk-heading-l
+    | Add business
+  h3.govuk-heading-m [class="govuk-!-margin-bottom-2"] Select business type
+= form_with(model: @business, local: true, url: wizard_path, method: :put) do |form|
+  = render "form_validation_display", errors: @business.errors
+  .govuk-grid-row class="govuk-!-margin-bottom-3"
+    .govuk-grid-column-one-half class="govuk-!-margin-bottom-9"
+      fieldset.govuk-fieldset
+        .govuk-form-group
+          .govuk-radios.govuk-radios--conditional data-module="radios"
```
Comment (broder) [inline mspsds-web/app/views/investigations/businesses/type.html.slim]: We're using `form_components/govuk_radios` to generate radio buttons now. @jasiekmiko might be able to help you set it up.

File: mspsds-web/test/system/investigation_business_test.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    click_on "Continue"
+    assert_text "Please select a business type"
+  end
+
+  test "cannot allow business type other to be empty" do
+    select_business_type_other
+    assert_text "Please enter a business type"
+  end
+
+  test "should delete business" do
```
Comment (broder) [inline mspsds-web/test/system/investigation_business_test.rb]: Does this actually delete the business or does it just unlink it from the case? Should we deleting the business if there's no way to re-add it to a case? Might be a question for designers.

File: mspsds-web/app/controllers/investigations/businesses_controller.rb [hunk 3]
```diff
@@ -1,37 +1,49 @@
 class Investigations::BusinessesController < ApplicationController
+  include ActiveModel::Validations
```
[the remark this answers is not in the document]
Comment (broder) [reply mspsds-web/app/controllers/investigations/businesses_controller.rb]: Did you see this comment?

File: mspsds-web/app/controllers/investigations/businesses_controller.rb [hunk 4]
```diff
@@ -1,37 +1,49 @@
...
   include Shared::Web::CountriesHelper
+  include Wicked::Wizard
+  skip_before_action :setup_wizard, only: %i[remove unlink]
+  steps :type, :details
 
-  before_action :set_investigation
-  before_action :set_business, only: %i[link remove unlink]
-  before_action :create_business, only: %i[new create]
-  before_action :set_countries, only: %i[new create]
+  validates :name, presence: true
```
Comment (broder) [inline mspsds-web/app/controllers/investigations/businesses_controller.rb]: I'm not sure what this is doing but it's probably related to the `ActiveModel::Validations`

File: mspsds-web/app/views/investigations/businesses/details.html.slim [hunk 5]
```diff
@@ -0,0 +1,16 @@
+- content_for :page_title, "Add business to case"
+- content_for :after_header do
+  = link_to "Back to #{@investigation.pretty_description}", @investigation, class: "govuk-back-link"
+.investigation-business-page{ data-investigation-id=@investigation.id
+    data-business-ids=@investigation.business_ids.join(",") }
+  = render "investigation_heading", investigation: @investigation
+  h2.govuk-heading-l
```
[the remark this answers is not in the document]
Comment (broder) [reply mspsds-web/app/views/investigations/businesses/details.html.slim]: I think we'll still want to address this.

File: mspsds-web/app/views/investigations/businesses/details.html.slim [hunk 6]
```diff
@@ -0,0 +1,16 @@
+- content_for :page_title, "Add business to case"
+- content_for :after_header do
+  = link_to "Back to #{@investigation.pretty_description}", @investigation, class: "govuk-back-link"
+.investigation-business-page{ data-investigation-id=@investigation.id
```
Comment (broder) [inline mspsds-web/app/views/investigations/businesses/details.html.slim]: Is this used for anything anymore?

File: mspsds-web/test/system/investigation_business_test.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    click_on "Continue"
+    assert_text "Please select a business type"
+  end
+
+  test "cannot allow business type other to be empty" do
+    select_business_type_other
+    assert_text "Please enter a business type"
+  end
+
+  test "should unlink business" do
```
Comment (broder) [inline mspsds-web/test/system/investigation_business_test.rb]: Did you ask anyone about whether this should actually delete the business?

File: mspsds-web/app/controllers/investigations/businesses_controller.rb (hunk 4 above)
Comment (samlionheart) [reply mspsds-web/app/controllers/investigations/businesses_controller.rb]: Yes sorry, I did see it, just reverted a commit and missed adding it back. They're both not needed

File: mspsds-web/test/system/investigation_business_test.rb (hunk 7 above)
Comment (samlionheart) [reply mspsds-web/test/system/investigation_business_test.rb]: Yeah, I'm waiting on Ed H

File: mspsds-web/app/views/investigations/businesses/details.html.slim (hunk 5 above)
[the remark this answers is not in the document]
Comment (samlionheart) [reply mspsds-web/app/views/investigations/businesses/details.html.slim]: Confused, I've added that component within the form, what elements should be reordered. The tests are failing around this bit on travis: https://travis-ci.org/UKGovernmentBEIS/beis-mspsds/jobs/498271147#L771
but my localhost is fine, and the errors are working

[the remark this answers is not in the document]
Comment (jasiekmiko) [reply mspsds-web/app/views/investigations/businesses/details.html.slim]: @samlionheart, does [this explanation](https://safetyandstandards.slack.com/archives/GCT2PLPG9/p1551263607017400) help?

File: mspsds-web/app/views/investigations/businesses/details.html.slim [hunk 8]
```diff
@@ -1,16 +1,14 @@
 - content_for :page_title, "Add business to case"
 - content_for :after_header do
   = link_to "Back to #{@investigation.pretty_description}", @investigation, class: "govuk-back-link"
-
-.investigation-business-page
-  = form_with(model: @business, url: investigation_businesses_path(@investigation), local: true, method: :post,
-          html: { autocomplete: :off }) do |form|
-    .govuk-grid-row
-      .govuk-grid-column-two-thirds-from-desktop
+  = render "investigation_heading", investigation: @investigation
```
Comment (jasiekmiko) [inline mspsds-web/app/views/investigations/businesses/details.html.slim]: Looks like all of the content on this page is inside of the `- content_for :after_header` block, which shouldn't be the case - you want to indent out everything from this point froward

File: mspsds-web/app/views/investigations/businesses/type.html.slim [hunk 9]
```diff
@@ -0,0 +1,19 @@
+- content_for :page_title, "Add business to case"
+- content_for :after_header do
+  = link_to "Back to #{@investigation.pretty_description}", @investigation, class: "govuk-back-link"
+  = render "investigation_heading", investigation: @investigation
```
Comment (jasiekmiko) [inline mspsds-web/app/views/investigations/businesses/type.html.slim]: Same here, only the link back should be inside the `content_for`

File: mspsds-web/app/views/investigations/businesses/type.html.slim [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  h2.govuk-heading-l
+    | Add business
+  h3.govuk-heading-m [class="govuk-!-margin-bottom-2"] Select business type
+= form_with(model: @business, local: true, url: wizard_path, method: :put) do |form|
+  = render "form_components/govuk_error_summary", form: form
```
Comment (jasiekmiko) [inline mspsds-web/app/views/investigations/businesses/type.html.slim]: We should wrap the contents of the page in 
```
.govuk-row
  .govuk-column-two-thirds-from-business
```

File: mspsds-web/app/views/investigations/businesses/details.html.slim [hunk 11]
```diff
@@ -1,16 +1,14 @@
...
+  h2.govuk-heading-l
+    | Add business
+  .govuk-grid-row
+    .govuk-grid-column-one-half
```
Comment (jasiekmiko) [inline mspsds-web/app/views/investigations/businesses/details.html.slim]: I would expect this to be `.govuk-grid-column-two-thirds-from-desktop`, is there a specifc reason why that's changed? :)

File: mspsds-web/test/system/investigation_business_test.rb (hunk 7 above)
Comment (samlionheart) [reply mspsds-web/test/system/investigation_business_test.rb]: His response was that products and businesses should be removed entirely when removed from a case, thought there's no ticket for this yet

File: mspsds-web/app/views/investigations/businesses/details.html.slim [hunk 12]
```diff
@@ -1,14 +1,14 @@
...
-        = render "form_components/govuk_error_summary", form: form
-        h3.govuk-heading-m[class="govuk-!-margin-bottom-2"]
-          | Business details
-        = render "businesses/form", form: form, business: @business, countries: @countries
+= render "investigation_heading", investigation: @investigation
+h2.govuk-heading-l
+  | Add business
+.govuk-grid-row
+  .govuk-grid-column-two-thirds-from-desktop
+    = form_with(model: @business, url: wizard_path, method: :put,
```
Comment (broder) [inline mspsds-web/app/views/investigations/businesses/details.html.slim]: @jasiekmiko, do we need to reorder the elements as in https://github.com/UKGovernmentBEIS/beis-mspsds/blob/master/mspsds-web/app/views/investigations/tests/details.html.slim?

Comment (jasiekmiko) [reply mspsds-web/app/views/investigations/businesses/details.html.slim]: Yes please, the headings (both the "investigation_heading" and the h2) should be within the 2/3s column, under the error summary.

File: mspsds-web/app/views/investigations/businesses/type.html.slim [hunk 13]
```diff
@@ -0,0 +1,21 @@
+- content_for :page_title, "Add business to case"
+- content_for :after_header do
+  = link_to "Back to #{@investigation.pretty_description}", @investigation, class: "govuk-back-link"
+= render "investigation_heading", investigation: @investigation
+h2.govuk-heading-l
+  | Add business
+.govuk-row
+  .govuk-column-two-thirds-from-business
+    h3.govuk-heading-m [class="govuk-!-margin-bottom-2"] Select business type
+    = form_with(model: @business, local: true, url: wizard_path, method: :put) do |form|
```
Comment (broder) [inline mspsds-web/app/views/investigations/businesses/type.html.slim]: Does this partial need re-arranging too?

**Pull request merged**